### ISSUE TITLE
Recommendations for Grid Sample/Affine Grid/Displacement Fields/Optical Flow

### ISSUE THREAD

**Issue by bnehoran** [pytorch/pytorch#24870]
I compiled a list of recommendations (as well as outstanding issues needing attention) for improving the `grid_sample` and `affine_grid` functions (the functions that are used to form a spatial transformer module), to make them easier to use for a number of computer vision applications.
The list is not exhaustive, so feel free to help out by adding other known issues and/or recommendations.

Some of these are more quick-and-easy fixes that could be easily moved out of the way, while others require some more in-depth discussion and design. For the latter category, I would appreciate discussion and feedback from some of the existing stakeholders such as @ssnl, @fmassa, @zou3519, @ailzhang, @soumith, and others.

### Bugs and Issues
- [#23925] Fix the border gradient issue, which causes training to occasionally get wild results for grid points at image borders.
- [#24823] Fix handling of `NaN` and `inf` grid inputs, which causes segfaults and corrupted CUDA memory (also #19826, [forum/39837](https://discuss.pytorch.org/t/what-may-cause-runtime-error-in-grid-sample/39837)).
- [#24821] `affine_grid` does currently support generating 3D affine grids. Update the documentation to support it, and update the tests to test the functionality. (also [#24380](https://github.com/pytorch/pytorch/issues/24380#issuecomment-522294803))
- `affine_grid` error messages: Produce a more useful error message when the affine matrix has the wrong shape or dtype (#12362), or when it is not a tensor.
- [#5565] `bilinear` mode for 3D `grid_sample` should really be `trilinear`. Personally, I would prefer to have just a single mode name for `linear`, `bilinear`, and `trilinear`, but this would bring it in line with `interpolate()`.

### Support Optical Flow More Directly
_These changes would greatly simplify the implementation of convolutional networks to do Optical Flow Estimation and Image Registration tasks._
- [Fixed in #23923] Add `align_corners=False` option to make `grid_sample` resolution agnostic, which is important for predicting optical flow at less-than-full resolution (originally #20785)
- Allow option for sampling using only the residual displacement/flow. Eliminates the need to constantly add an identity grid to the flow/displacement field, which is imprecise, slow, and very [prone to user error](https://discuss.pytorch.org/t/warp-video-frame-from-optical-flow/6013).
- Allow option for the grid/flow input to be in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which, again, is prone to user error, (especially when users don’t know which setting of `align_corners` to target). 

Implementation: Combinations of these options (residual and/or pixel units) could be added to `grid_sample()`, or as suggested in #4085, put in a separate `flow_sample()` high-level function (which should probably use the same kernels as the existing `grid_sample()`). Either way, the kernel code needs to be changed very little in order to allow them. For pixel units, just skip the call to the `unnormalize` function on the grid. For residual displacement, just add the destination x,y coordinates to the (unnormalized) grid.

### Inconsistencies and Conventions
_These inconsistencies are not necessarily incorrect, so to speak, but some of these do tend to cause a lot of headaches and confusion._
- Change padding modes names in `grid_sample` to match the names of padding modes in [`torch.nn.functional.pad`](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.pad).
- The function which on the user-facing python side is named `grid_sample`, is behind the scenes on both the python and the C++ side always named `grid_sampler` (notice the extra ‘r’) and `GridSampler`. Perhaps standardize it by renaming the behind-the-scenes versions to `grid_sample` (respectively `GridSample`), because this direction would not break backwards compatibility. Could still cause BC issue with saved models, though.
- [Fixed in #23923] Add `align_corners` option so that `grid_sample` matches the indexing conventions of `interpolate` (originally #20785)
- Switch from a “channels-last” to a “channels-first” convention in order to match the format of all other PyTorch tensors. That is, `[N, C, H, W]` rather than `[N, H, W, C]`, where `C` represents the grid components. Would eliminate the need to constantly call `permute()` to move channels back and forth. This would allow grids/flows/fields to be used normally in most other PyTorch functions, such as `interpolate`, among others. (The only other exceptions to the channels-first rule, as far as I know, are the FFT functions, which also expect channels-last `[N, H, W, C]`.)
Similarly, the components should be expressed as `(y,x)` instead of the current `(x,y)` to match the indexing order. This is currently very confusing for [users](https://twitter.com/ducha_aiki/status/1024596593302011904). 
Downside: would be very BC-breaking.
- Standardize support for half precision grids in CUDA kernels. This was partially implemented before, but in a way that caused problems for double precision (by casting everything to single precision floats). Perhaps there should be a conditional cast only if the input grid has less than single precision.

### New Features
- [#2732] Support broadcasting in `grid_sample` and `affine_grid`.
Specifically, for `affine_grid`, allow it to accept a single 2x3 2D affine matrix, or a single 3x4 3D affine matrix, with no batch dimension, and just produce a single-batch grid (or one with no batch dimension in conjunction with `grid_sample` broadcasting).
For `grid_sample`, apply broadcasting rules to the batch dimension. (Broadcasting doesn’t make sense in the spatial dimensions, of course.)
- Allow `affine_grid` to accept a python list or a numpy `ndarray`, since when not used as part of a spatial transformer module, users often think of `affine_grid` as a constructor. For example, construct a grid tensor that rotates an image. Can possibly just have it call `torch.tensor()` on the input if not already a tensor.
- Add a cyclic/circular padding mode, where the image is treated as if repeated/tiled in each dimension
- Add option for bicubic interpolation. This would be similar to the bicubic mode of of `interpolate()` (#9849, #19703).
Perhaps also add a [lanczos](https://en.wikipedia.org/wiki/Lanczos_resampling) mode like [Torch has](https://github.com/torch/image/blob/master/doc/paramtransform.md).
- [#21457] @fmassa has suggested adding the capability for each grid point to sample from a region defined by its neighboring grid points, mirroring the `area` mode of interpolate, rather than just a 2-pixel neighborhood in each dimension. It’s not immediately clear to me how to define this region and the interpolation method, so this requires some research.
If designed and implemented properly, this could subsume the `interpolate()` method’s functionality for downsampling when using an identity grid of smaller size. I wouldn’t recommend using it in place of `interpolate`, since it would probably be much less performant and require generating an affine grid, but this could help in building a better Spatial Transformer module that minimizes aliasing.

### Refactoring
- There is a ‘dispatcher’ for `affine_grid_generator` that is now left all alone in `nn/_functions/vision.py`, and its sole purpose is to decide if to use cuDNN. I don’t see why we need to keep `vision.py` around just for that. It could probably be moved either up directly into the `affine_grid` function of `nn/functional.py` or down into the C++ implementation.
- Profile CUDA kernels vs. cuDNN kernels, and consider dropping use of cuDNN grid sampler and affine grid generator if they don’t give much extra value, as [suggested](https://github.com/pytorch/pytorch/issues/20785#issuecomment-496014108) by @soumith
- Vectorize the 3D CPU kernel of `grid_sample`. See if gives performance gain like it does for the 2D version at #10980.
- Group together some of the code for `grid_sample` and `interpolate`. They have some shared behavior and kernel code (such as the bilinear/trilinear interpolation) that could potentially be standardized and factored out.

**Comment by ailzhang** [comment]
Fyi I moved one item above to be #24470 instead of "need discussion". ;) That should be totally fine as long as the perf gap is not large. ;)

**Comment by andreaskoepf** [comment]
Regarding `grid_sample` and `interpolate`: How is down sampling currently implemented in `interpolate`? As far as I know downsampling of an image requires low-pass filtering the source image prior sampling. All frequencies that cannot be represented at the target sampling-resolution would ideally have to be removed. 

AFAIK `grid_sample` only looks at the direct neighborhood of the source image. This gives reasonable values between the discrete source samples (pixels) and reconstruct a continuous function (what you want during up-sampling, sub-pixel shifting, rotation etc.) but it does not low-pass filtering by dynamically widening the filter support window.

I just took a look at the imagemagick resize function (they had a nice [page about resampling filters](http://www.imagemagick.org/Usage/filter/)) and they seem to widen the filter support for the down sampling case, e.g. see: [resize.c#L3296-L3297](https://github.com/ImageMagick/ImageMagick/blob/0a403f460a5fc0855d53f4b640e6e2f66f5be94c/MagickCore/resize.c#L3296-L3297) - if the target image is half the size of the source image the filter-window width would be doubled and the filter function (which generates the coefficients) itself is also 'stretched'. I also found this 1990 paper [Filters for Common Resampling Tasks](http://www.realitypixels.com/turk/computergraphics/ResamplingFilters.pdf) which shows different filters (box, tent, gaussian, lanczos) for down-sampling (decimation).

I had no not yet time to drill down to the PyTorch interpolation code but I noticed that the functions all have an 'upsample_' prefix. Is there a special case implemented for down-sampling? The flow-field that is passed to `grid_sample` can potentially present very strange mappings... Maybe a filter window-width or scaling factor could be passed here to support the down-scaling scenario.

**Comment by bnehoran** [comment]
@andreaskoepf would you be interested in helping to implement some of these sampling filters?
We have an issue open at #21457 specifically for discussing how to allow `grid_sample` to better simulate downsampling. Part of the challenge there is exactly what you mention, that `grid_sample` currently only interpolates from the immediate nearest neighborhood, and in order to get it to downsample properly, you need it to have a much wider window/receptive field.

I'm pretty sure the functions named `upsample_`-something (the ones that are dispatched to by `interpolate`) actually handle both cases of upsampling and downsampling, although, as with `grid_sample` they don't do as well on downsampling (for the same reason). By contrast, the `area` mode dispatches to `adaptive_avg_pool2d`, which downsamples by average pooling. Of course, a similar mode for `grid_sample` is non-trivial because of the irregular grid.

**Comment by fmassa** [comment]
Hi @bnehoran 

Great issue !

One more thing I'd like to add to the `interpolate` / `grid_sample` similarities. For `interpolate` (and probably `grid_sample` as well), we could probably factor out the kernel implementations so that they are separable, meaning that the temporal / spatial / volumetric versions all dispatch to the same kernel, greatly simplifying the code.

While making it separable is kind of easy, it would still be great to keep the same performance as before, so avoiding materializing the intermediate results would be better I think.

Related issue: https://github.com/pytorch/pytorch/issues/10482

**Comment by ailzhang** [comment]
Hi @bnehoran , thanks for the great proposal! Here're decisions came out of our meeting:

```
* Fix the border gradient issue, which causes training to occasionally get wild results for grid points at image borders.


[DECISION]: Yes we should definitely fix it, this is high priority. https://quip.com/PfVAEArlMDl would you mind if I assign this to you so that you take a look when you have bandwidth? 


* Allow option for sampling using only the residual displacement/flow. Eliminates the need to constantly add an identity grid to the flow/displacement field, which is imprecise, slow, and very prone to user error (https://discuss.pytorch.org/t/warp-video-frame-from-optical-flow/6013).

[DECISION]: Yes we should do it, it’s a good option to have 


* Allow option for the grid/flow input to be in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which, again, is prone to user error,

[DECISION]: Yes absolute coordinate is great, we want this.  And for residual displacement, we can have a new flow_sample() high level API. 



* The function which on the user-facing python side is named grid_sample, is behind the scenes on both the python and the C++ side always named grid_sampler (notice the extra ‘r’) and GridSampler. Perhaps standardize it by renaming the behind-the-scenes versions to grid_sample (respectively GridSample), because this direction would not break backwards compatibility. Could still cause BC issue with saved models, though.


[DECISION]: Yes cpp side should be changed 



* Switch from a “channels-last” to a “channels-first” convention in order to match the format of all other PyTorch tensors. That is, [N, C, H, W] rather than [N, H, W, C], where C represents the grid components. Would eliminate the need to constantly call permute() to move channels back and forth. This would allow grids/flows/fields to be used normally in most other PyTorch functions, such as interpolate, among others. (The only other exceptions to the channels-first rule, as far as I know, are the FFT functions, which also expect channels-last [N, H, W, C].)


[DECISION]: maybe not necessary for now, we can decide later.



* Standardize support for half precision grids in CUDA kernels of grid_sample. This was partially implemented before, but in a way that caused problems for double precision (by casting everything to single precision floats). Perhaps there should be a conditional cast only if the input grid has less than single precision.

[DECISION]: Yes we want this.


* Allow affine_grid to accept a python list or a numpy ndarray, since when not used as part of a spatial transformer module, users often think of affine_grid as a constructor. For example, construct a grid tensor that rotates an image. Can possibly just have it call torch.tensor() on the input if not already a tensor.

[DECISION]: Yes this sounds fine 


* Add new interpolation modes, including bicubic, area, and possibly lanczos (https://en.wikipedia.org/wiki/Lanczos_resampling).

[DECISION] : Yes this is great to have



* Improve the ability of grid_sample to downsample while warping (that is, when the grid points are spaced out far apart). This would allow a single warp-downsample operation which is more capable than applying the operations in sequence.

[DECISION]: this needs some more research 



* Group together some of the code for grid_sample and interpolate. They have some shared behavior and kernel code (such as the bilinear/trilinear interpolation) that could potentially be standardized and factored out.

[DECISION]: Yes we should refactor interpolate to be more generic, but it’ll require some effort 
```

We can safely remove the [Discussion] label on the approved items. :D

**Comment by ailzhang** [comment]
Contributions to this list are very well welcome, if you are interested in helping, please drop a comment here and we're happy to help getting started.

**Comment by riccardosamperna** [comment]
> * Allow option for sampling using only the residual displacement/flow. Eliminates the need to constantly add an identity grid to the flow/displacement field, which is imprecise, slow, and very prone to user error (https://discuss.pytorch.org/t/warp-video-frame-from-optical-flow/6013).
> 
> [DECISION]: Yes we should do it, it’s a good option to have 
> 
> 
> * Allow option for the grid/flow input to be in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which, again, is prone to user error,

I am not sure this is the right place to ask so please remove the comment if not appropriate. 
I have a question about this two points highlighted above. I am currently working on a image registration algorithm and as pointed out in one of the points above I have a displacement field. Can you please explain me which is the right way to use the grid_sample function. I explain below what I mean with some pseudocode:

```
Option 1: 
grid = meshgrid(linspace(-1, 1, width), linspace(-1, 1, height))
final_grid = grid + displacement_field
final_grid = normalize(final_grid) # normalize in range [-1, 1] using (width, height)

Option 2:
grid = meshgrid(linspace(-1, 1, width), linspace(-1, 1, height))
dispalcement_field = normalize(displacement_field) # normalize in range [-1, 1] using (width, height)
final_grid = grid + displacement_field
```

**Comment by ismarou** [comment]
@riccardosamperna I believe you have to choose the first option as the grid_sample documentation page states that the final_grid should lie between -1 and 1, but I am not sure either, as most implementations found after a bit of searching select the second option.

**Comment by tychovdo** [comment]
> ```
> Option 1: 
> grid = meshgrid(linspace(-1, 1, width), linspace(-1, 1, height))
> final_grid = grid + displacement_field
> final_grid = normalize(final_grid) # normalize in range [-1, 1] using (width, height)
> 
> Option 2:
> grid = meshgrid(linspace(-1, 1, width), linspace(-1, 1, height))
> dispalcement_field = normalize(displacement_field) # normalize in range [-1, 1] using (width, height)
> final_grid = grid + displacement_field
> ```

@riccardosamperna, be careful here. As @bnehoran mentioned the conversion between absolute coordinates and normalized coordinates can be tricky and is prone to user error. I believe both of these options are such errors, since the grid and displacement field are not added in world coordinate space. It's probably best to first transform/create a grid in world coordinates (opposed to [-1, 1] space), then add your displacement field and then normalize the added result to obtain a [-1, 1] input required for `F.grid_sample`.

Maybe this can help:
https://github.com/voxelmorph/voxelmorph/blob/7f289b3613bdbccfdb4c1eff3fbff29710a65680/pytorch/model.py#L168

**Comment by riccardosamperna** [comment]
Thanks for the explanation and the pointer @tychovdo. I have been trying to wrap my head around this topic for a long time now and in most of the projects in Github using the grid_sample I still see a lot of wrong implementations. It looks like that I am not the only one struggling with this F.grid_sample documentation/interface.

**Comment by bnehoran** [comment]
@riccardosamperna That's a good question, and I'd be happy to help you, but I think the more appropriate place to discuss this would be the PyTorch forum (https://discuss.pytorch.org).
If you ask this question in the forum and ping me, I'd be happy to help you out there.

Comments on this issue should ideally be more related to implementation details and directions rather than usage.

**Comment by xieshuqin** [comment]
@bnehoran  It would be great to include higher order derivative for this operation too. Computing second order derivatives is quite common in meta-learning settings but grid_sample doesn't support this yet.

**Comment by BaldrLector** [comment]
Hi @bnehoran

Great issue !

There one more problem in grid_sample.
report by here, some method could train well in v1.0.x, but model diverge if you use Pytorch 1.1 and higher.
reported in
[https://github.com/AliaksandrSiarohin/first-order-model](first order)
[https://github.com/AliaksandrSiarohin/first-order-model/issues/13](url)
[https://github.com/AliaksandrSiarohin/first-order-model/issues/4](url)

**Comment by SomeoneDaniel** [comment]
Is it possible to use grid_sample to mimic bilinear_resize with half_pixel=false?

**Comment by PeterL1n** [comment]
I opened a new suggestion for `grid_sample`: #63016.

Basically, `grid_sample` should not require `input` and `grid` to have the same dimension. For a 4D input `[B, C, H, W]`, grid like `[B, D, 2]` should also work.

**Comment by xming521** [comment]
Anything new?
